src/main/java/org/bytedeco/javacpp/Loader.java

```diff
@@ -771,14 +789,43 @@ public static File extractResource(URL resourceURL, File directoryOrFile,
                             file.delete();
                             String s = resourceURL.toString();
                             URL u = new URL(s.substring(0, s.indexOf("!/") + 2) + entryName);
+                            // FIXME: check if directories have to be extracted again?!
                             file = extractResource(u, file, prefix, suffix);
                         }
                         file.setLastModified(entryTimestamp);
                     }
                 }
+                System.out.println("Extract took " + (System.currentTimeMillis() - start) + "ms, for:" + resourceURL);
                 return directoryOrFile;
             }
         }
+        if (OSGiBundleResourceLoader.isOSGiRuntime()) {
+            Enumeration<URL> directoryEntries = OSGiBundleResourceLoader.getBundleDirectoryContent(resourceURL);
+            if (directoryEntries != null && directoryEntries.hasMoreElements()) { // a not empty directory
+                String directoryName = resourceURL.getPath();
+                while (directoryEntries.hasMoreElements()) {
```

Comment:
This while loop looks pretty similar to the one above. Is the idea that some OSGi implementations may support other things that JAR files?

Reply:
This is intentionally. Actually I would have liked to move the extraction part into an own method. But because in case of a JarURLConnection the URL is created just before extractResource is called again this is not possible without creating the URL before (and therefore also in cases of a directory.

And yes a bundle can be a jar but also can be extracted into a directory (which is for example very common if you develop Eclipse plug-ins and use them in a debugged application started from the IDE).